src/coreComponents/constitutive/capillaryPressure/CapillaryPressureBase.cpp

```diff
@@ -26,22 +26,31 @@ using namespace dataRepository;
 namespace constitutive
 {
 
-constexpr integer CapillaryPressureBase::PhaseType::GAS;
-constexpr integer CapillaryPressureBase::PhaseType::OIL;
-constexpr integer CapillaryPressureBase::PhaseType::WATER;
-
 namespace
 {
 
-std::unordered_map< string, integer > const phaseDict =
+integer toPhaseType( string const & lookup, string const & groupName )
 {
-  { "gas", CapillaryPressureBase::PhaseType::GAS   },
-  { "oil", CapillaryPressureBase::PhaseType::OIL   },
-  { "water", CapillaryPressureBase::PhaseType::WATER }
-};
-
+  static std::unordered_map< string, integer > const phaseDict =
+  {
+    { "gas", CapillaryPressureBase::PhaseType::GAS },
+    { "oil", CapillaryPressureBase::PhaseType::OIL },
+    { "water", CapillaryPressureBase::PhaseType::WATER }
+  };
+  auto const it = phaseDict.find( lookup );
+  if( it == phaseDict.end() )
+  {
+    std::vector< string > phaseNames;
+    std::transform( phaseDict.begin(), phaseDict.end(), std::back_inserter( phaseNames ), []( auto const & p ){ return p.first; } );
```

Comment:
Yeah, performance is not a concern (and is equivalent in this case). Expressiveness is actually my main goal when I try to use standard algorithms. Unfortunately, current C++ and classic STL make it a little too verbose with iterators. 

I'd argue that with ranges (C++20), this
```c++
auto phaseNames = phaseDict | std::views::transform( []( auto && p ){ return p.first; } );
```
spells out "extract keys from the dict" much cleaner than the loop version, and C++ devs should get used to more functional style of programming :smile: 
(this also has performance benefits, as it constructs a lazily evaluated range that you can pass directly to `join` or `std::format` or some other function that concatenates the values, and avoid allocating the vector).

But back to reality... yeah, I guess I don't mind either way.

Reply:
Yes, `ranges` seems to be great.

I would argue that it's maybe an element that would bring a big big booster to GEOSX codebase.
We use extensively this kind of methods
```c++
class Group {
...
  template< typename GROUPTYPE = Group, typename ... GROUPTYPES, typename LAMBDA >
  void forSubGroups( LAMBDA && lambda ) const
...
}
```
With a predicate (only based on type by design) and with a callback/functor pattern, at the noteworthy difference w.r.t the STL that the dev using these functions does not own the container on which it operates. Making the chaining of these calls more complicated. We sometimes end in nested lambda and nested captures... 

Also, in some cas... Oh, I realize I was going to repeat https://github.com/GEOSX/GEOSX/issues/1089!

There exists a c++14 implementation of the `ranges` lib https://github.com/ericniebler/range-v3.
_I'm convinced that this could literally change GEOSX for the best._
So we may want to include it. That would be a smart move imho.

> C++ devs should get used to more functional style of programming 😄

This is actually not *only* a functional pattern issue.
It's just that the boilerplate that C++ brings is not worth the price here 😉 

Let's assume that you want to filter a list and apply a function on it.
In `python` you have
```python
map( f, filter( p, lst ) )
```
In `java` (and maybe `javascript` with the same syntax)
```java
stream.filter( p ).map( f )
```
in `common lisp` something like
```lisp
( mapcar f ( remove-if-not p lst ) )
```
You are young, you're surely brave enough to tell me how to do this in std C++ without `ranges`!